scipy/stats/tests/test_multivariate.py

```diff
@@ -1024,6 +1026,68 @@ def test_wishart_invwishart_2D_rvs(self):
         assert_allclose(frozen_iw_rvs, manual_iw_rvs)
 
 
+class TestSpecialOrthoGroup(TestCase):
+    def test_reproducibility(self):
+        np.random.seed(514)
+        x = special_ortho_group.rvs(3)
+        expected = np.array([[ 0.99394515, -0.04527879,  0.10011432],
+                             [-0.04821555,  0.63900322,  0.76769144],
+                             [-0.09873351, -0.76787024,  0.63295101]])
+        assert_array_almost_equal(x, expected)
+
+    def test_frozen_matrix(self):
+        dim = 7
+        frozen = special_ortho_group(dim)
+
+        rvs1 = frozen.rvs(random_state=1234)
+        rvs2 = special_ortho_group.rvs(dim, random_state=1234)
+
+        assert_equal(rvs1, rvs2)
+
+    def test_det_and_ortho(self):
+        xs = [ special_ortho_group.rvs(dim)
+               for dim in range(2,12)
+               for i in range(3)
+             ]
+        # Test that determinants are always +1
+        dets = [ np.linalg.det(x) for x in xs ]
+        assert_allclose(dets, [1.]*30)
+
+        # Test that these are orthogonal matrices
+        for x in xs:
+            assert_array_almost_equal(np.dot(x, x.T),
+                                      np.eye(x.shape[0]))
+
+    def test_haar(self):
+        # Test that the distribution is constant under rotation
+        # Every column should have the same distribution
+        # Additionally, the distribution should be invariant under another rotation
+
+        # Generate samples
+        dim = 5
+        samples = 500 # Not too many, or the test takes too long
+        decimal_tol = 0 # ...so don't expect much precision
+        xs = [ special_ortho_group.rvs(dim) for i in range(samples) ]
+
+        # Dot a few rows (0, 1, 2) with unit vectors (0, 2, 4, 3),
+        #   effectively picking off entries in the matrices of xs.
+        #   These projections should all have the same disribution,
+        #     establishing rotational invariance.
+        #   We could instead test that angles between random vectors
+        #     are uniformly distributed, but the below is sufficient.
+        #   It is not feasible to consider all pairs, so pick a few.
+        els = ( (0,0), (0,2), (1,4), (2,3)  )
+        proj = { (er, ec): [ x[er][ec] for x in xs ] for er, ec in els }
+        pairs = [ (e0, e1) for e0 in els for e1 in els if e0 > e1 ]
+        for (p0, p1) in pairs:
+            # regress proj[p0] against proj[p1]
+            # These should be equally distributed, so
+            #   should get a straight line, b=0, m=1, X2 not too big
+            slope, intercept, corr_coeff, j, sterrest = linregress(sorted(proj[p0]), sorted(proj[p1]))
+            assert_almost_equal(slope, 1, decimal=decimal_tol)
+            assert_almost_equal(intercept, 0, decimal=decimal_tol)
+            assert_almost_equal(sterrest, 0, decimal=decimal_tol)
```

Comment:
How long is the test taking? I was given a budget of 1s and I'm using it up on my machine. I can increase the number of samples, but otherwise I was hitting the tolerance too often.

In addition to the timing budget, do we have a rule of thumb for failing frequency in tests like this? We can't guarantee to pass them every time; how often can we fail?